Use relative path in installed libraries

fbudin69500 opened the pull request:
On MacOS, libraries can be modified at installation to update the path to shared libraries used using the tool `install_name_tool`. On linux, there is no easy way to fixup the paths to the libraries that are used (`patchelf` can only update the library paths in its most recent version that is not widely available). Instead, on Linux, the relative path in which libraries should be found at runtime is added to RPath. This patch allows to not manually have to set (DY)LD_LIBRARY_PATH when using `pydrake`: All the dependencies will be found automatically.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7301)
<!-- Reviewable:end -->

fbudin69500 commented:
`chrpath` on linux could work. It would add a dependency, and the logic that has been added to `pybind.bzl` would move to `install.py.in`(it would be similar to what is done for MacOS). Both approaches should give the exact same result. It would not allow to change the individual path of each library like `install_name_tool` or `patchelf` would.

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-Kwq2IpDuHWsaFIiUtTI:buxngc8)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/bindings/pybind.bzl, line 27 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8HSS50kIc_db30gPn:-Kx8HSS50kIc_db30gPo:b-4m7nhp) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/drake/bindings/pybind.bzl#L27)):*
> ```Python
> 
>     # Count folders in name to find relative path to lib subfolder.
>     relative_path = '/..' * (name.count('/') + 1)
> ```

It is unclear why the count of two slashes found in `"python/pydrake/_pydrake_common.so"` are directly related to the number of `..` steps we need to go from, e.g., `"lib/python2.7/site-packages/pydrake/_pydrake_symbolic.so"` to `"lib/libdrake.so"`.

Probably the way to make it clear would be to assert that `name.startswith("python/")`, then count that as two (explaining `"python#.#/site-packages"`), and then take any additional module paths (one, in most cases) and count those as well.

---

*[tools/install.py.in, line 101 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8GV3vxC13GTFH9oNj:-Kx8GV3vxC13GTFH9oNk:brjezn4) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L101)):*
> ```
>                  dst_full]
>                 )
>         # Remove rpath values that contain $ORIGIN. These are from the build
> ```

The `fix_library_paths` is only run in the case of `sys.platform == "darwin"`.  The changes in `pybind.bzl` only add `"-Wl,-rpath,$$ORIGIN" ...` when not `"//tools/cc_toolchain:apple"`.  Thus, I don't understand where `$ORIGIN` is coming from.  Is Bazel itself adding that rpath, outside of our skylark code?

---

*[tools/install.py.in, line 106 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8FFiXL6PxJI4uqicK:-Kx8FFiYw2uQg6ljwS5x:b1yby4v) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L106)):*
> ```
>         for line in file_output.splitlines():
>             split_line = line.strip().split(' ')
>             if len(split_line) >= 2 and 'path' == split_line[0] and split_line[1].startswith('$ORIGIN'):
> ```

line too long

---

*[tools/install.py.in, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8FPbRjUUzvTQ5JrOy:-Kx8FPbRjUUzvTQ5JrOz:b15sd25) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L107)):*
> ```
>             split_line = line.strip().split(' ')
>             if len(split_line) >= 2 and 'path' == split_line[0] and split_line[1].startswith('$ORIGIN'):
>                 # path is specified 2 lines below
> ```

I don't understand what this comment means.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/bindings/pybind.bzl, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8KlBGwVcK7pyWdQT2:-Kx8KlBHE8QBSlQ6orpy:b74bj69) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/drake/bindings/pybind.bzl#L26)):*
> ```Python
>             fail("%s cannot be set by the caller" % key)
> 
>     # Count folders in name to find relative path to lib subfolder.
> ```

To avoid confusing a reader of the code here, this documentation here should make it clear that we are performing an installation-related action at build-time instead of install-time, because we do not want to depend on `chrpath` during install.  In other words, make it clear that this `rpath` is not relevant for the sandbox nor for building nor testing, but rather _only_ once this `*.so` file has been copied into an install tree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Marking high priority since this is required for one of the tests in #7293 to pass.

fbudin69500 commented:
Review status: 1 of 3 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[tools/install.py.in, line 101 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8GV3vxC13GTFH9oNj:-Kx8pLJMAmYYVSp4-8BR:bom1crx) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L101)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The `fix_library_paths` is only run in the case of `sys.platform == "darwin"`.  The changes in `pybind.bzl` only add `"-Wl,-rpath,$$ORIGIN" ...` when not `"//tools/cc_toolchain:apple"`.  Thus, I don't understand where `$ORIGIN` is coming from.  Is Bazel itself adding that rpath, outside of our skylark code?
</blockquote></details>

To answer the question, the linker adds values to RPATH to find the libraries in the build tree. This is done both on linux and on MacOS, and not limited to bazel. On MacOS, this can create the SIP error reported by @ EricCousineau-TRI in https://github.com/RobotLocomotion/drake/issues/6760 . Since the paths added to RPATH at build time are irrelevant at install time, it is better to remove them.
However, in this new version of the PR, the fix is now done at install time for both MacOS and Linux. This is because the previous version of the patch was only addressing the issue for the pydrake libraries, whereas this new patch should solve the problem for all the libraries (one limitation is that the libraries must already contain an RPATH, otherwise they cannot be fixed at install time, e.g. VTK libraries).

---

*[tools/install.py.in, line 106 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8FFiXL6PxJI4uqicK:-Kx8tLWcb4RGoqDPTImZ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L106)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

line too long
</blockquote></details>

Done.

---

*[tools/install.py.in, line 107 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8FPbRjUUzvTQ5JrOy:-Kx8tgu5busZeMD_e3qQ:bss4a2b) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L107)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I don't understand what this comment means.
</blockquote></details>

Removed.

---

*[drake/bindings/pybind.bzl, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-Kx8KlBGwVcK7pyWdQT2:-KxATXM1wtxxq0nyEvS1:b-od9wdv) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/drake/bindings/pybind.bzl#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

To avoid confusing a reader of the code here, this documentation here should make it clear that we are performing an installation-related action at build-time instead of install-time, because we do not want to depend on `chrpath` during install.  In other words, make it clear that this `rpath` is not relevant for the sandbox nor for building nor testing, but rather _only_ once this `*.so` file has been copied into an install tree.
</blockquote></details>

This new version of the patch actually removes the modifications of `pybind.bzl` and uses `chrpath` to update RPATH on linux at install time. This is necessary to fix all the libraries and not only `pydrake` libraries.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 3 files at r2.
Review status: 1 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
One problem that this patch does not address is the RPATH of the executables, but that doesn't seem to be needed for now.

---

Review status: 1 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxAV6oOcRV8qYtDwU6z:b-g3dcao)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
PR needs a rebase onto recent master.

---

Reviewed 2 of 3 files at r2.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAWCs7RqXALAZzxhRq:-KxAWCs8N7RKDzJTZodd:bfjww0i) (no related file):*
We're past the point where the amount of code in `install.py.in` is consistent with a codegen template (a `*.in` file).  We're missing out on pycodestyle checking of it, as well as the easy ability to unit test the intricate new (and existing) functions. 

Ideally, all of the library fixup methods would have standalone branch-coverage level unit testing with python's unittest library that include both common and corner cases.  It's great that we have `testCommonInstall.py` as an acceptance test, but focused unit testing would enable reviewers to rely on CI, instead of having to put the code through its paces locally, or rely on out-of-tree CI to test it.

Those function need to turn into `*.py` code that is imported by the generated `:install` program(s).  Since this PR is high priority, I would be willing to accept a high-priority issue number that will be fixed immediately after this PR merges, instead of blocking this PR on it.

---

*[tools/install.py.in, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzMFk4dcH3m_I15h2:-KxAzMFk4dcH3m_I15h3:b2fec11) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L68)):*
> ```
>         libs[basename] = (dst_full, installed)
> ```

BTW Needs two lines between function declarations per pycodestyle.

---

*[tools/install.py.in, line 17 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAV_53EHTY2peyJ-Mw:-KxAV_53EHTY2peyJ-Mx:b5rzu8a) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L17)):*
> ```
> # On linux, dynamic libraries may have their version number
> # as a suffix (e.g. my_lib.so.x.y.z).
> dylib_match = r'(.*.so)(\.\d+)*$|(.*dylib)$'
> ```

BTW Do you want to escape the `.` in `.so` to force it to literally match a period?

---

*[tools/install.py.in, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAVrhQWdIz0vlEL_YI:-KxAVrhQWdIz0vlEL_YJ:bjrx73p) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L60)):*
> ```
>         print("[Up to date] %s" % dst)
>     m = re.match(dylib_match , dst)
>     if m != None:
> ```

FYI Python styleguides usually say `m is not None` as the required spelling.

---

*[tools/install.py.in, line 85 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAW2iVtQnZu6iefkHh:-KxAW2iVtQnZu6iefkHi:b5hkkj2) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L85)):*
> ```
> 
> ```

BTW nit Three blank lines?  Should be two, I think.

---

*[tools/install.py.in, line 122 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzZ9yXnAGnXs0j8SV:-KxAzZ9yXnAGnXs0j8SW:b2fec11) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L122)):*
> ```
>             ['install_name_tool', "-delete_rpath", split_line[1], dst_full]
>             )
> ```

BTW Needs two lines between function declarations per pycodestyle.

---

*[tools/install.py.in, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzbuFinsgVXZ22HxN:-KxAzbuGBTK61RJd_swX:bexg8nh) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L126)):*
> ```
>     # Check that library has an rpath or a runpath tag
>     try:
>       check_output(["chrpath", "-l", dst_full])
> ```

Bad indentation.  Run a Python style checker, I guess manually for now.

---

*[tools/install.py.in, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzuErKGsME4VfPCqK:-KxAzuEs-MYJREqGZVcP:b-vojwd9) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L126)):*
> ```
>     # Check that library has an rpath or a runpath tag
>     try:
>       check_output(["chrpath", "-l", dst_full])
> ```

FYI Consider printing an informative error message if `chrpath` program is not present.  (Since it's a new dependency, some users are likely to forget to install it.)

---

*[tools/install.py.in, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzlBdzVGzhEpNeMgo:-KxAzlBdzVGzhEpNeMgp:b5bplk9) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L129)):*
> ```
>     except CalledProcessError as ex:
>       if ex.output.strip().endswith('no rpath or runpath tag found.'):
>         # Cannot be modified with `chrpath`, so we skip it.
> ```

If the `no rpath` message is not present in the output we ... still ignore the error?  Shouldn't we re-raise it instead, or print and exit, or ...?

---

*[tools/install.py.in, line 139 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-Lgj5NjOcA2-WO3G:-KxB-Lgj5NjOcA2-WO3H:bqor9yy) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L139)):*
> ```
>         if len(ldd_result) < 2 or ldd_result[1] != "not found":
>             continue
>         m = re.match(dylib_match, ldd_result[0])
> ```

Python best practice would be to to a structured assignment, something like:
```python
soname, sover = m.groups()
```
In this way, if the number of groups in the regex ever evolves, we would fail-fast instead of computing with lies.

---

*[tools/install.py.in, line 148 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-kBVd_FBxQ1M1qcM:-KxB-kBVd_FBxQ1M1qcN:birvyz7) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L148)):*
> ```
>           os.path.relpath(libs[m.group(1)][0], lib_dirname)
>         )
>         if diff_path:
> ```

Unclear when this is ever false?  Should the `/` just appear in the subsequent line, instead?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxAy_x3ZHofMTnNvXww:bxm1z1o)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 0 of 3 files reviewed at latest revision, 11 unresolved discussions.

---

*[tools/install.py.in, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzMFk4dcH3m_I15h2:-KxDcpbs82T1HqLNfbiB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/87e47798789c6a87d5bc65d738c6679bb3df6b04/tools/install.py.in#L68)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Needs two lines between function declarations per pycodestyle.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 17 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAV_53EHTY2peyJ-Mw:-KxDc-tr-ur3LRyxF1tZ:b86r3ei) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L17)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Do you want to escape the `.` in `.so` to force it to literally match a period?
</blockquote></details>

Good idea, and same with dylib.

---

*[tools/install.py.in, line 60 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAVrhQWdIz0vlEL_YI:-KxDcIQVg7Ax4UnZgjYF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L60)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Python styleguides usually say `m is not None` as the required spelling.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 85 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAW2iVtQnZu6iefkHh:-KxDcqya3nWTVhuh8FVz:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L85)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW nit Three blank lines?  Should be two, I think.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 122 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzZ9yXnAGnXs0j8SV:-KxDcrlq0-yvSyrSwd3x:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L122)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Needs two lines between function declarations per pycodestyle.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzbuFinsgVXZ22HxN:-KxDkm8os2OFUbmo6HFB:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L126)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Bad indentation.  Run a Python style checker, I guess manually for now.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 126 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzuErKGsME4VfPCqK:-KxEWrG92CMS_8fOxz86:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L126)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider printing an informative error message if `chrpath` program is not present.  (Since it's a new dependency, some users are likely to forget to install it.)
</blockquote></details>

Done.

---

*[tools/install.py.in, line 129 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxAzlBdzVGzhEpNeMgo:-KxEXFKJO5Tz3La7Ap7D:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L129)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If the `no rpath` message is not present in the output we ... still ignore the error?  Shouldn't we re-raise it instead, or print and exit, or ...?
</blockquote></details>

Done.

---

*[tools/install.py.in, line 139 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-Lgj5NjOcA2-WO3G:-KxEWnNZSHTWeKSJJgAW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L139)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Python best practice would be to to a structured assignment, something like:
```python
soname, sover = m.groups()
```
In this way, if the number of groups in the regex ever evolves, we would fail-fast instead of computing with lies.
</blockquote></details>

Done.

---

*[tools/install.py.in, line 148 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-kBVd_FBxQ1M1qcM:-KxEXuF9t1IFWl3Vui_F:b-al0bjj) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L148)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear when this is ever false?  Should the `/` just appear in the subsequent line, instead?
</blockquote></details>

It was to avoid having a trailing `/` in `RPATH`, butit does not matter.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 4 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[tools/install/install.py.in, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEZj3eGEG6W0COOLCT:-KxEZj3eGEG6W0COOLCU:b-3qum0t) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L138)):*
> ```
>         if ex.errno == 2 and ex.strerror == "No such file or directory":
>             print("`chrpath` not found. Please run install_prereqs.sh.")
>             sys.exit(1)
> ```

The not-if condition should re-raise, I think?
```python
    except OSError as ex:
        if ex.errno == 2 and ex.strerror == "No such file or directory":
            print("`chrpath` not found. Please run install_prereqs.sh.")
        raise ex
    except ...
```

---

*[tools/install/install.py.in, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxE_0BBGqJ265RevOk7:-KxE_0BBGqJ265RevOk8:b-xflk63) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L144)):*
> ```
>                 # Cannot be modified with `chrpath`, so we skip it.
>                 return
>         else:
> ```

When returncode is 2, but the output didn't match, we sliently ignore the error. 

Probably the `else` should disappear, and the re-raise should de-indent four spaces?

---

*[tools/install.py.in, line 148 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-kBVd_FBxQ1M1qcM:-KxE_L_ZNvHsZ_34kUBu:buk5le5) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L148)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

It was to avoid having a trailing `/` in `RPATH`, butit does not matter.
</blockquote></details>

OK But I didn't think diff_path would ever be empty?  It would be `.`, at worst.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[tools/install/install.py.in, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEZj3eGEG6W0COOLCT:-KxElXFpBCJW3phljtDE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L138)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The not-if condition should re-raise, I think?
```python
    except OSError as ex:
        if ex.errno == 2 and ex.strerror == "No such file or directory":
            print("`chrpath` not found. Please run install_prereqs.sh.")
        raise ex
    except ...
```
</blockquote></details>

Done.

---

*[tools/install/install.py.in, line 144 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxE_0BBGqJ265RevOk7:-KxEnDzwKOMhoxxcXDIC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L144)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

When returncode is 2, but the output didn't match, we sliently ignore the error. 

Probably the `else` should disappear, and the re-raise should de-indent four spaces?
</blockquote></details>

Done.

---

*[tools/install.py.in, line 148 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-kBVd_FBxQ1M1qcM:-KxEjv1y8ByyVGx_PtV2:bha10lf) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L148)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

OK But I didn't think diff_path would ever be empty?  It would be `.`, at worst.
</blockquote></details>

Actually it does return an empty string. `os.path.relpath()` returns `.` but `os.path.dirname()` returns an empty string when called on `.`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 4 files at r3.
Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[tools/install/install.py.in, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEZj3eGEG6W0COOLCT:-KxEy__2FczwNLMGWseM:bomc5n5) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L138)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think the re-raise should be out one level of indentation?  So that it always happens?

(This is why PRs without unit tests are a challenge, for both author and reviewer.)

---

*[tools/install/install.py.in, line 143 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEzzQ_YYThy9kY8jbh:-KxEzzQ_YYThy9kY8jbi:b63je9j) ([raw file](https://github.com/robotlocomotion/drake/blob/71bc83b92345c8e4b79e3bff1c3f0842a33b5996/tools/install/install.py.in#L143)):*
> ```
>             and ex.output.strip().endswith('no rpath or runpath tag found.'):
>                 # Cannot be modified with `chrpath`, so we skip it.
>                 return
> ```

Wrong level of intendation -- should be only +4 versus the if this body is filling in.

---

*[tools/install.py.in, line 148 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxB-kBVd_FBxQ1M1qcM:-KxEwahk3Cr0JZcVZs-E:bajz6iu) ([raw file](https://github.com/robotlocomotion/drake/blob/37aadefbbee561ed26953d1d730af274aac8e6aa/tools/install.py.in#L148)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

Actually it does return an empty string. `os.path.relpath()` returns `.` but `os.path.dirname()` returns an empty string when called on `.`
</blockquote></details>

OK I see now.  If you wanted to put it back with a comment, I'd be OK with that too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/install/install.py.in, line 138 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEZj3eGEG6W0COOLCT:-KxF2q0A4Fg2YbUfXjXS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/680477d09504d0785b26dca1e6321b74fa7f1a97/tools/install/install.py.in#L138)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think the re-raise should be out one level of indentation?  So that it always happens?

(This is why PRs without unit tests are a challenge, for both author and reviewer.)
</blockquote></details>

Done.

---

*[tools/install/install.py.in, line 143 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxEzzQ_YYThy9kY8jbh:-KxF6Y1D6u9Ykz7FnYKL:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/71bc83b92345c8e4b79e3bff1c3f0842a33b5996/tools/install/install.py.in#L143)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Wrong level of intendation -- should be only +4 versus the if this body is filling in.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging) The commits bled into each other a bit.  FYI Just doing a squash in GitHub merge process here is OK by me.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> module the one open discussion below (filing an issue).

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxF9IYOwKwK8PSZHsFh:b-9w0g8b)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
-(status: curate commits before merging) 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxFC6n7co8mV-DuZEJs:b-l70wzf)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
+@sammy-tri Platform review please

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxFC5pKOkTQbe9EUWdY:b-xvgex6)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tools/install/install.py.in, line 162 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxFNYpokoTu-iN33387:-KxFNYpokoTu-iN33388:b52rehw) ([raw file](https://github.com/robotlocomotion/drake/blob/9f0095fabf07a79c4823e9be1efc7e0cc4307c95/tools/install/install.py.in#L162)):*
> ```
>             os.path.relpath(libs[soname][0], lib_dirname)
>         )
>         rpath.add('$ORIGIN' + '/' + diff_path)
> ```

I'm confused here...  Are we deleting rpath entries with `$ORIGIN` in them in the mac version, and then adding rpaths with `$ORIGIN` in the linux version?  I think a couple of comments describing a typical "before" and "after" rpath in each case might be helpful.  (I've got a decent grasp of ELF dynamic linking but there's a lot of context at this point of execution of the script which I'm not familiar with, sorry... )

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
Review status: 2 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/install/install.py.in, line 162 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxFNYpokoTu-iN33387:-KxJRrvZUtAFxaGULbRj:b-32nmee) ([raw file](https://github.com/robotlocomotion/drake/blob/9f0095fabf07a79c4823e9be1efc7e0cc4307c95/tools/install/install.py.in#L162)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

I'm confused here...  Are we deleting rpath entries with `$ORIGIN` in them in the mac version, and then adding rpaths with `$ORIGIN` in the linux version?  I think a couple of comments describing a typical "before" and "after" rpath in each case might be helpful.  (I've got a decent grasp of ELF dynamic linking but there's a lot of context at this point of execution of the script which I'm not familiar with, sorry... )
</blockquote></details>

On MacOS, we do not use RPATH as the path to each library is individually set (relative or absolute depending if it is part of the package or a system library). On linux, libraries are found using RPATH. The RPATH values used in the build tree are removed with the `check_output` call below while the new RPATH values corresponding to the install tree are written. Let me know if the new comments are clearer.

---

*[tools/install/install.py.in, line 156 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxJSNfyTHxb1lzJL0No:-KxJSNfyTHxb1lzJL0Np:b6qpw5b) ([raw file](https://github.com/robotlocomotion/drake/blob/e86f56e3a0b9d214b6ab54f2e7c629fba41e82e4/tools/install/install.py.in#L156)):*
> ```
>         if len(ldd_result) < 2 or \
>                 not (ldd_result[1] == "not found"
>                      or ldd_result[1].startswith(prefix)):
> ```

Added a test here to make sure that any library that is has a path containing the `prefix` also has its relative path to the current target also updated, just in case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[tools/install/install.py.in, line 162 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-KxFNYpokoTu-iN33387:-KxJztuxdr7_q9_Z_EQp:b-obcaym) ([raw file](https://github.com/robotlocomotion/drake/blob/9f0095fabf07a79c4823e9be1efc7e0cc4307c95/tools/install/install.py.in#L162)):*
<details><summary><i>Previously, fbudin69500 (Francois Budin) wrote…</i></summary><blockquote>

On MacOS, we do not use RPATH as the path to each library is individually set (relative or absolute depending if it is part of the package or a system library). On linux, libraries are found using RPATH. The RPATH values used in the build tree are removed with the `check_output` call below while the new RPATH values corresponding to the install tree are written. Let me know if the new comments are clearer.
</blockquote></details>

I think that's a bit clearer, thanks

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxK-34KnxaI6dKxFdhx:bnfp4nl)*
<!-- Sent from Reviewable.io -->

fbudin69500 commented:
I filed an issue for the lack of testing for `install.py.in` : https://github.com/RobotLocomotion/drake/issues/7331

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxK4xiyJifD_i8YTofK:b-ndnqzr)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 4 files at r3, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7301#-:-KxK5bL8p2Mm8Lhr5jXI:bnfp4nl)*
<!-- Sent from Reviewable.io -->
